From: Egbert Eich <eich@suse.de>
Date: Sat Aug 22 10:30:50 2015 +0200
Subject: drm/i915: Queue reenable timer also when enable_hotplug_processing is false
Patch-mainline: N/A
References: bsc#942938

If an HPD IRQ storm happens before everything is fully set up, the
reenable timer will not be started which causes HPD to be turned off
until reboot.
To avoid this process the HPD state even when enable_hotplug_processing is
not set.

Signed-off-by: Egbert Eich <eich@suse.de>
Signed-off-by: Egbert Eich <eich@suse.com>
---
 drivers/gpu/drm/i915/i915_irq.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)
diff --git a/drivers/gpu/drm/i915/i915_irq.c b/drivers/gpu/drm/i915/i915_irq.c
index 22e23bb..391eb45 100644
--- a/drivers/gpu/drm/i915/i915_irq.c
+++ b/drivers/gpu/drm/i915/i915_irq.c
@@ -699,10 +699,6 @@ static void i915_hotplug_work_func(struct work_struct *work)
 	bool changed = false;
 	u32 hpd_event_bits;
 
-	/* HPD irq before everything is fully set up. */
-	if (!dev_priv->enable_hotplug_processing)
-		return;
-
 	mutex_lock(&mode_config->mutex);
 	DRM_DEBUG_KMS("running encoder hotplug functions\n");
 
@@ -740,6 +736,12 @@ static void i915_hotplug_work_func(struct work_struct *work)
 
 	spin_unlock_irqrestore(&dev_priv->irq_lock, irqflags);
 
+	/* HPD irq before everything is fully set up. */
+	if (!dev_priv->enable_hotplug_processing) {
+		mutex_unlock(&mode_config->mutex);
+		return;
+	}
+
 	list_for_each_entry(connector, &mode_config->connector_list, head) {
 		intel_connector = to_intel_connector(connector);
 		intel_encoder = intel_connector->encoder;
